libs/langchain/langchain/embeddings/base.py

```diff
@@ -1,4 +1,226 @@
+        )
+    if not model:
+        raise ValueError("Model name cannot be empty")
+    return provider, model
+
+
+def _infer_model_and_provider(
+    model: str, *, provider: Optional[str] = None
+) -> Tuple[str, str]:
+    if not model.strip():
+        raise ValueError("Model name cannot be empty")
+    if provider is None and ":" in model:
+        provider, model_name = _parse_model_string(model)
+    else:
+        provider = provider
+        model_name = model
+
+    if not provider:
+        providers = sorted(_SUPPORTED_PROVIDERS)
+        raise ValueError(
+            "Must specify either:\n"
+            "1. A model string in format 'provider:model-name'\n"
+            "   Example: 'openai:text-embedding-3-small'\n"
+            "2. Or explicitly set model_provider from: "
+            f"{providers}"
+        )
+
+    if provider not in _SUPPORTED_PROVIDERS:
+        raise ValueError(
+            f"Provider '{provider}' is not supported.\n"
+            f"Supported providers and their required packages:\n"
+            f"{_get_provider_list()}"
+        )
+    return provider, model_name
+
+
+def _init_embedding_model_helper(
+    model: str, *, provider: Optional[str] = None, **kwargs: Any
+) -> Embeddings:
+    """Initialize an Embeddings model from the model name and provider.
+
+    Internal helper function that handles the actual model initialization.
+    Use init_embedding_model() instead of calling this directly.
+    """
+    provider, model_name = _infer_model_and_provider(model, provider=provider)
+    pkg = _SUPPORTED_PROVIDERS[provider]
+    _check_pkg(pkg)
+
+    if provider == "openai":
+        from langchain_openai import OpenAIEmbeddings
+
+        return OpenAIEmbeddings(model=model_name, **kwargs)
+    elif provider == "azure_openai":
+        from langchain_openai import AzureOpenAIEmbeddings
+
+        return AzureOpenAIEmbeddings(model=model_name, **kwargs)
+    elif provider == "google_vertexai":
+        from langchain_google_vertexai import VertexAIEmbeddings
+
+        return VertexAIEmbeddings(model=model_name, **kwargs)
+    elif provider == "bedrock":
+        from langchain_aws import BedrockEmbeddings
+
+        return BedrockEmbeddings(model_id=model_name, **kwargs)
+    elif provider == "cohere":
+        from langchain_cohere import CohereEmbeddings
+
+        return CohereEmbeddings(model=model_name, **kwargs)
+    elif provider == "mistralai":
+        from langchain_mistralai import MistralAIEmbeddings
+
+        return MistralAIEmbeddings(model=model_name, **kwargs)
+    elif provider == "huggingface":
+        from langchain_huggingface import HuggingFaceEmbeddings
+
+        return HuggingFaceEmbeddings(model_name=model_name, **kwargs)
+    else:
+        raise ValueError(
+            f"Provider '{provider}' is not supported.\n"
+            f"Supported providers and their required packages:\n"
+            f"{_get_provider_list()}"
+        )
+
+
+@functools.lru_cache
```

Comment:
should we set max size

Reply:
sure! though I hope people don't spam the provider list